### PR TITLE
Implement sin and cos for decimal32

### PR DESCRIPTION
**Pull request by mborland**
Closes: #71 
Closes: #89

### CODE REVIEW

**Comment by mborland** [comment]
@ckormanyos now that the constants are in this is the first of the transcendental functions I was going to hit. Do you think this is an appropriate amount of terms in the polynomial? Or is there a different way you would go? Hopefully constructing all the constants at compile time helps a bit, but the repeated `fma` is going to be expensive.

**Comment by ckormanyos** [comment]
> an appropriate amount of terms in the polynomial? Or is there a different way you would go?

We have less than ten decimal digits of precisoin. But I would maybe try to reduce the argument even further with angle reduction to something around less than 1/4. Then maybe do a Pade expansion using about 5 terms.

When you've reduced the argument, go for something like this at [Wolfram Alpha](https://www.wolframalpha.com/input?i=FullSimplify%5BPadeApproximant%5BSin%5Bx%5D%2C+%7Bx%2C+0%2C+%7B5%2C4%7D%7D%5D%5D)

In other words:

```
FullSimplify[PadeApproximant[Sin[x], {x, 0, {5,4}}]]
```

There are also some great, classic references (which I posess) on this kind of thing.

[1] J.F. Hart, _Computer_ _Approximations_, Robert E. Krieger Publishing 1968, 1978.
[2] W.J. Cody and W. Waite, _Software_ _Manual_ _for_ _the_ _Elementary_ _Functions_, Prentice Hall 1980

**Review by ckormanyos — APPROVED** [review]
This looks really nice. It is very simple and terse. I do note that we could be really close to a generic implementation if the coefficient table(s) were to be specialized. But we can cross that bridge when we get to it.

**Comment by mborland** [comment]
> This looks really nice. It is very simple and terse. I do note that we could be really close to a generic implementation if the coefficient table(s) were to be specialized. But we can cross that bridge when we get to it.

Now that `remquo` should be complete I am going to try and reduce the angle to 45 deg and dispatch between `sin` and `cos`. The coefficients have enough precision for `decimal64`, but If/when we get to dec128 that will be its own set of fun.

**Comment by ckormanyos** [comment]
Hi Matt (@mborland) I might have gotten a bit click happy here and I merged develop into the sin-branch and handled merge conflict in a test file. I'm pretty sure you would have wanted to do this yourself. So that was my mistake. Sorry. If I messed up big-time, I'll help repair.

**Comment by mborland** [comment]
> Hi Matt (@mborland) I might have gotten a bit click happy here and I merged develop into the sin-branch and handled merge conflict in a test file. I'm pretty sure you would have wanted to do this yourself. So that was my mistake. Sorry. If I messed up big-time, I'll help repair.

Don't worry about it. I am sure we will have more conflicts as we both work on <cmath> in parallel.

File: include/boost/decimal/detail/cmath/sin.hpp [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+namespace boost { namespace decimal {
+
+template<typename T, std::enable_if_t<detail::is_decimal_floating_point_v<T>, bool> = true>
+constexpr auto sin(T x) noexcept -> T
+{
+    constexpr T zero {0, 0};
+
+    // First check non-finite values
+    if (abs(x) == zero || isinf(x) || isnan(x))
```
[the remark this answers is not in the document]
Comment (mborland) [reply include/boost/decimal/detail/cmath/sin.hpp:24]: That definitely makes more sense

**Comment by ckormanyos** [comment]
>  try and reduce the angle to 45 deg and dispatch between `sin` and `cos`

I remember in the old days we would make `sincos()`. I don't know if it ever got into C/C++ but it was convenient for `tan`, and all kinds of stuff like FFTs, computations with recurrence relations and what not. But if it's not in the language it's hard to find the _right_ _place_ for `sincos`, ... sigh...
